---
title: "Rev-Erba in Th17 RNA-seq Practice"
output: github_document
author: Jiayi Dong
---
This is a demo analysis for the RNA-seq data from Amir et al., 2018, Cell Reports 25, 3733â€“3749.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

In this work, we use REV-ERBa WT versus KO RNA-seq from Th17 cells [GEO: GSE122726](https://ftp.ncbi.nlm.nih.gov/geo/series/GSE122nnn/GSE122726/suppl/GSE122726_processed_218.xlsx)
There're two biological replicates for each genotypes in this experiment.


### 1. Load `DESeq2` package for differential gene expresion analysis.
```{r}
library(DESeq2)
library(dplyr)
```


### 2. Prepare data object for analysis 
#### 2.1 DESeq2 Required Inputs
As input, the DESeq2 package expects **(1)** a data.frame of **count data** (as obtained from RNA-seq or another high-throughput sequencing experiment) and **(2)** a second data.frame with information about the samples - often called sample metadata (or `colData` in DESeq2-speak because it supplies metadata/information about the columns of the countData matrix). 

#### 2.2 Load count data from GEO database
**Note**: The count values must be raw counts of sequencing reads. This is important for **DESeq2**'s statistic model to hold.

```{r}
# load count table 
count<- read.csv("GSE122726 WTVSKO.csv",stringsAsFactors = FALSE,row.names = 1)
rownames(count) <- gsub("\t", "", rownames(count))
head(count)
```

Some genes have **zero** read counts for all 4 samples. We tidy up the counts data to exclude genes with **zero** value.
```{r}
filter.Count <- count[rowSums(count)!=0,]
head(filter.Count)
```

#### 2.3 Create colData
Create a `meta` for metadata
```{r}
id <- colnames(filter.Count)

meta <- data.frame("id"=id[1:4], "condition" = c(rep("WT",2),rep("KO",2)), stringsAsFactors = FALSE)
```

Take a look at metadata
```{r}
head(meta)
```

### 3. DESeq2 analysis
We will use the **DESeqDataSetFromMatrix()** function to build the required DESeqDataSet object and call it `dds`, short for our DESeqDataSet. 

```{r}
dds <- DESeqDataSetFromMatrix(countData=filter.Count, 
                              colData=meta, 
                              design=~condition
                              )
#set WT as reference level
dds$condition <- relevel(dds$condition, ref = "WT")

```

We're ready to run the differential expression pipeline. We use **DEseq()** to run DESeq2 analysis.
```{r}
dds <- DESeq(dds)
head(dds)
```

### 4. Inspect result tale
Get Deseq results by calling **results()** function
```{r}
res <- results(dds)
summary(res)
```

#### 4.1 Dignostic plot

##### MA-plot

The MA plot represents each gene with a dot. The x axis is the average expression over all samples, the y axis the log2 fold change between treatment and control. Genes with an adjusted p value below a treshold (0.1 by default) are shown in red.
```{r}
# unshrunken
plotMA(res,ylim=c(-2,2))
```

The MA-plot shows the log2 fold changes from the treatment over the mean of normalized counts, i.e. the average of counts normalized by size factor. 


When count values are too low to allow an accurate estimate of the log fold changes(LFCs), the value is "shrunken" towards zero to avoid that these values dominate the top-ranked log fold changes.
```{r}
#Shrunken
res_shrink <- lfcShrink(dds, contrast=c("condition","KO","WT"), res=res)
plotMA(res_shrink,ylim=c(-2,2))
```

##### Plot of despersion estimates

The function **plotDispEsts** visualizes DESeq2's dispersion estimates.


```{r}
# Plot dispersion estimates
plotDispEsts(dds, main="DESeq2")
```
The black points are the dispersion estimates for each gene.We fit the red trend line, and shrink each gene's estimate towards the red line to obtain the final estimates(blue points) that are used in the hypothesis test. The blue circles above the main "cloud" of points are genes which have high gene-wise dispersion estimates which are labelled as dispersion outliers.

##### Plot counts

This is to examine the counts of reads for a single gene across the groups. `plotCounts()` normalizes counts by sequencing depth and adds a pseudocount of 1/2 to allow for log scale plotting. 
```{r}
plotCounts(dds, gene=which.min(res$padj),intgroup = "condition")
```

##### Blind dispersion estimation
```{r}
rld <- rlog(dds,blind=FALSE)
vsd <- varianceStabilizingTransformation(dds,blind=FALSE)
vsd.fast <- vst(dds,blind=FALSE)
head(assay(rld),3)
head(assay(vsd),3)
head(assay(vsd.fast),3)

```

##### Sample distances

Assess overall similarity between samples is a useful first step in an RNA-Seq analysis.

Here we plot sample distances.
```{r}
sampleDists <- dist(t(assay(rld)))
```

```{r}
library("pheatmap")
library("RColorBrewer")
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(rld$condition,rld$id)
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette(rev(brewer.pal(9,"Blues")))(100)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col=colors)
```

```{r}
#plot p-value distribution
#hist(resAdjSig05$pvalue)
```

##### Heatmap of the count matrix

```{r}
select <- order(rowMeans(counts(dds,normalized=TRUE)),decreasing =TRUE)[1:20]
nt <- normTransform(dds) #defaults to log2(x+1)
log2.norm.counts <- assay(nt)[select,]
df <- data.frame("condition" = c(rep("WT",2),rep("KO",2)))
rownames(df) <- meta$id
pheatmap(log2.norm.counts,cluster_rows = FALSE,show_rownames = FALSE,cluster_cols = FALSE,annotation_col =df)
pheatmap(assay(rld)[select,],cluster_rows = FALSE,show_rownames = FALSE,cluster_cols = FALSE,annotation_col =df)
pheatmap(assay(vsd)[select,],cluster_rows = FALSE,show_rownames = FALSE,cluster_cols = FALSE,annotation_col =df)

```

##### PCA plot
```{r}
plotPCA(rld,intgroup = "condition")
```


#### 4.2 Getting results of differentially expressed genes


Subset the DEG with adjust p value lower than 0.05
```{r}
resAdjSig05 <- subset(as.data.frame(res),padj <0.05)
head(resAdjSig05)
```

Get upregulated and downregulated gene, respectively
```{r}
resAdjSig05up <- subset(resAdjSig05,resAdjSig05$log2FoldChange>0)
resAdjSig05down <- subset(resAdjSig05,resAdjSig05$log2FoldChange<0)
head(resAdjSig05up)
head(resAdjSig05down)
```

Save result with DE gene ordered by padj
```{r}
write.csv(resAdjSig05, "adjsignif05_results.csv")
write.csv(resAdjSig05up, "adjsignif05_results_up.csv")
write.csv(resAdjSig05down, "adjsignif05_results_down.csv")

```

##### Genome annotation 
```{r}
library(org.Mm.eg.db)
library("AnnotationDbi")
```

```{r}
columns(org.Mm.eg.db)
```
```{r}
res$entrez <-unname(mapIds(org.Mm.eg.db, rownames(res), "ENTREZID", "SYMBOL"))
res$ensembl <- unname(mapIds(org.Mm.eg.db, rownames(res), "ENSEMBL", "SYMBOL"))
```
```{r}
head(res)
```
```{r}
resOrdered <- res[order(res$pvalue),]
head(resOrdered)
```

Save results
```{r}
write.csv(as.data.frame(resOrdered), file="results.csv")
```

#### 4.3 Data visualization


##### Heatmap for gene clustering
```{r}
library("genefilter")
topVarGenes <- head(order(rowVars(assay(rld)), decreasing = TRUE), 100)

```

```{r}
library("gplots")
```


```{r}
#head(rownames(resOrdered))
LFCvsdup <- subset(assay(vsd), rownames(assay(vsd)) %in% rownames(resAdjSig05up))
LFCvsddown <- subset(assay(vsd), rownames(assay(vsd)) %in% rownames(resAdjSig05down))


LFCtopVarGenesup <- head(order(rowVars(LFCvsdup),decreasing=TRUE),40)
LFCtopVarGenesdown <- head(order(rowVars(LFCvsddown), decreasing = TRUE),67)

#Heatmap for topVarGenes from resAdjSig05up
heatmap.2(LFCvsdup[LFCtopVarGenesup,], scale = "row", trace = "none", dendrogram = "column",
col = colorRampPalette(rev(brewer.pal(9, "RdBu")))(255))

heatmap.2(LFCvsddown[LFCtopVarGenesdown,], scale = "row", trace = "none", dendrogram = "column",Colv=FALSE,
col = colorRampPalette(rev(brewer.pal(9, "RdBu")))(255),labRow = rownames(LFCvsddown)[LFCtopVarGenesdown])
```

```{r}
pheatmap(LFCvsdup[LFCtopVarGenesup,],scale = "row",clustering_distance_rows = "correlation",cluster_rows = FALSE,show_rownames = TRUE,cluster_cols = FALSE,annotation_col =df)
```



##### GSEA and pathway analysis

GSEA with `gage` package
```{r}
library(gage)
library(pathview)

```

```{r}
library(gageData)
data("kegg.sets.mm")
data(sigmet.idx.mm)
```

```{r}
kegg.sets.M=kegg.sets.mm[sigmet.idx.mm]
head(kegg.sets.mm,3)
```

```{r}
foldchanges <-  res$log2FoldChange
names(foldchanges) <- res$entrez
head(foldchanges)
```



```{r}
keggres = gage(foldchanges,gsets = kegg.sets.mm)
attributes(keggres)
```
```{r}
keggl <- keggres$less
keggs <- keggres$stats
```

```{r}
lapply(keggres,head)
```
```{r}
keggrespathways <- data.frame(id=rownames(keggres$greater,keggres$greater %>% tbl_df() %>% filter(row_number()<=5 %>% .$id %>% as.character())))
head(keggrespathways,20)
```


```{r}
keggresids <- substr(keggrespathways$id,start=1,stop=8)
head(keggresids)
```

??
```{r}
#plot_pathway <- function(pid) pathview(gene.data = foldchanges,pathway.id = pid, species = "mm")
#tmp <- sapply(keggresids,function(pid) pathview(gene.data = foldchanges,pathway.id = pid,species="mm"))
```


## GO analysis

```{r}
data("go.sets.mm")
data("go.subs.mm")
gobpsets <- go.sets.mm[go.subs.mm$BP]
gobpres <- gage(foldchanges,gsets = gobpsets,same.dir = TRUE)
lapply(gobpres,head,10)
```

```{r}
gobprespathways <- data.frame(id=rownames(gobpres$greater,gobpres$greater %>% tbl_df() %>% filter(row_number()<=5 %>% .$id %>% as.character())))
head(gobprespathways,20)

```

```{r}
gobpresterm <- substr(gobprespathways$id,12,100)
head(gobpresterm,20)
```



```{r}
#keggrld <- assay(rld)
#head(keggrld)
#geneData(genes = keggresids, exprs = assay(rld)[], outname = outname,txt=T, heatmap = T,limit = 3,scartterplot=TRUE)
```

```{r}
#gobprespathways$id[1]
```


```{r}
#for (gs in gobprespathways$greater){
  #outname=gsub(" |:|/","_",substr(gs,12,100))
  #geneData(genes = kegg.sets.mm[[gs]], exprs = assay(rld), outname=outname,txt=T,heatmap=T,limit = 3,scatterplot = T)
#}

```


```{r}
#assay(rld)
#keggresids
```

